lib/immutable-struct.rb

```diff
@@ -67,9 +85,37 @@ def self.new(*attributes,&block)
     end
     klass.class_exec(&block) unless block.nil?
     imethods = klass.instance_methods(include_super=false)
-    klass.class_exec(imethods) do |imethods|
+    klass.class_exec(imethods, attributes) do |imethods, attributes|
+
+      derived_methods = imethods - attributes - [:to_h]  #to_h is never a derived method, despite its 0-arity
+
+      ##
+      # :method: attributes_to_h
+      # Return a +Hash+ with only those attributes defined in the class constructor.
+      define_method(:attributes_to_h) do
+        attributes.each_with_object({}) do |attr, hash|
+          ivar = attr.to_s.prepend(?@).to_sym
+          hash[attr] = self.instance_variable_get(ivar)
+        end
+      end
+
+      ##
+      # :method: derived_to_h
+      # Return a +Hash+ of the results all zero-arity methods that are not attributes
+      # defined in the class constructor 
+      define_method(:derived_to_h) do
+        derived_methods.each_with_object({}) do |derived_method, hash|
+          hash[derived_method] = self.send(derived_method) if 0 == self.method(derived_method).arity
+        end
+      end
+
+      alias_method :__to_h__, :to_h if method_defined?(:to_h)
+      ##
+      # :method: to_h
+      # Return a +Hash+ with the result of merging +attributes_to_h+ and +derived_to_h+.
+      # This method can also be overriden so that, for example, only +attributes_to_h+ is used.
       define_method(:to_h) do
-        imethods.inject({}){ |hash, method| hash.merge(method.to_sym => self.send(method)) }
+        self.respond_to?(:__to_h__) ? __to_h__ : attributes_to_h.merge(derived_to_h)
```

Comment:
Un saludo,
//nando
https://github.com/nandosola

----
"It is a mistake to think you can solve any major problems just with potatoes"
- Douglas A

> On 15 Jun 2015, at 20:23, David Copeland <notifications@github.com> wrote:
> 
> In lib/immutable-struct.rb:
> 
> >        define_method(:to_h) do
> > -        imethods.inject({}){ |hash, method| hash.merge(method.to_sym => self.send(method)) }
> > +        self.respond_to?(:__to_h__) ? __to_h__ : attributes_to_h.merge(derived_to_h)
> Sorry I'm having a hard time understanding this. What happens if we remove the alias and just have this method merge the two helper methods' hashes? i.e. what goes wrong that the __to_h__ method is solving?
> 
> —
> Reply to this email directly or view it on GitHub.
>